### PR TITLE
feat: add monitor on table chain rule set setelem and obj events

### PR DESCRIPTION
**Pull request by singchia**
- add monitor feature
- optimize multiple header types assertion(which needed by monitor)

### CODE REVIEW

File: monitor.go [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+			}
+		}
+	}
+	monitor.mtx.Lock()
+	if monitor.status != monitorClosed {
+		monitor.status = monitorClosed
+		monitor.closer()
+		close(monitor.eventCh)
+	}
+	monitor.mtx.Unlock()
```
Comment (stapelberg) [inline monitor.go:262]: Always call Unlock() via defer: 

```go
monitor.mu.Lock()
defer monitor.mu.Unlock()
```

File: monitor.go [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+	monitor.conn = conn
+	monitor.closer = closer
+
+	if monitor.monitorFlags != 0 {
+		err = conn.JoinGroup(uint32(unix.NFNLGRP_NFTABLES))
+		if err != nil {
+			monitor.closer()
+			return nil, err
+		}
+		conn.JoinGroup(uint32(unix.NFNLGRP_NFTRACE))
```
Comment (stapelberg) [inline monitor.go:291]: Why are we ignoring the error here?

File: monitor.go [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+func (cc *Conn) AddMonitor(monitor *Monitor) (chan *Event, error) {
+	conn, closer, err := cc.netlinkConn()
+	if err != nil {
+		return nil, err
+	}
+	monitor.conn = conn
+	monitor.closer = closer
+
+	if monitor.monitorFlags != 0 {
+		err = conn.JoinGroup(uint32(unix.NFNLGRP_NFTABLES))
```
Comment (stapelberg) [inline monitor.go:286]: if err := conn.JoinGroup(uint32(unix.NFNLGRP_NFTABLES)); err != nil {

**Review by stapelberg — CHANGES_REQUESTED** [review]
Thanks for your PR! Did a first review, but I probably have more comments once these are out of the way :)

File: monitor.go [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+	MonitorObjectAny  MonitorObject = MonitorObjectMask
+)
+
+var (
+	monitorFlags         map[MonitorAction]map[MonitorObject]uint32
+	monitorFlagsInitOnce sync.Once
+)
+
+// A lazy init function to define flags.
+func lazyInit() {
```
[the remark this answers is not in the document]
Comment (singchia) [reply monitor.go:54]: ok, learned

File: monitor.go [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+)
+
+// A Monitor to track actions on objects.
+type Monitor struct {
+	action       MonitorAction
+	object       MonitorObject
+	monitorFlags uint32
+
+	// mtx covers eventCh and status
+	mtx     sync.Mutex
```
[the remark this answers is not in the document]
Comment (singchia) [reply monitor.go:135]: sure

File: monitor.go [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+}
+
+func (monitor *Monitor) monitor() {
+	for {
+		msgs, err := monitor.conn.Receive()
+		if err != nil {
+			break
+		}
+		for _, msg := range msgs {
+			if got, want := msg.Header.Type&0xff00>>8, netlink.HeaderType(unix.NFNL_SUBSYS_NFTABLES); got != want {
```
[the remark this answers is not in the document]
Comment (singchia) [reply monitor.go:196]: ok

File: monitor.go (hunk 1 above)
Comment (singchia) [reply monitor.go]: ok fixed

File: monitor.go [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+	EventTypeDelSet     EventType = unix.NFT_MSG_DELSET
+	EventTypeNewSetElem EventType = unix.NFT_MSG_NEWSETELEM
+	EventTypeDelSetElem EventType = unix.NFT_MSG_DELSETELEM
+	EventTypeNewObj     EventType = unix.NFT_MSG_NEWOBJ
+	EventTypeDelObj     EventType = unix.NFT_MSG_DELOBJ
+)
+
+type Event struct {
+	Type  EventType
+	Data  interface{}
```
[the remark this answers is not in the document]
Comment (stapelberg) [reply monitor.go:112]: Oh, now I see why you couldn’t use `any` :) Yes, you can bump it to 1.18.

File: monitor.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+	MonitorObjectRuleset
+	MonitorObjectMask MonitorObject = (1 << iota) - 1
+	MonitorObjectAny  MonitorObject = MonitorObjectMask
+)
+
+var (
+	monitorFlags = map[MonitorAction]map[MonitorObject]uint32{
+		MonitorActionAny: {
+			MonitorObjectAny:      0xffffffff,
+			MonitorObjectTables:   1<<unix.NFT_MSG_NEWTABLE | 1<<unix.NFT_MSG_DELCHAIN,
```
Comment (stapelberg) [inline monitor.go:52]: This line says NEWTABLE and DELCHAIN. Shouldn’t it be NEWTABLE and DELTABLE?

Same in the copy in line 58.

**Review by stapelberg — CHANGES_REQUESTED** [review]

File: monitor.go [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+	EventTypeNewSet     EventType = unix.NFT_MSG_NEWSET
+	EventTypeDelSet     EventType = unix.NFT_MSG_DELSET
+	EventTypeNewSetElem EventType = unix.NFT_MSG_NEWSETELEM
+	EventTypeDelSetElem EventType = unix.NFT_MSG_DELSETELEM
+	EventTypeNewObj     EventType = unix.NFT_MSG_NEWOBJ
+	EventTypeDelObj     EventType = unix.NFT_MSG_DELOBJ
+	EventTypeOOB        EventType = math.MaxInt
+)
+
+type Event struct {
```
Comment (stapelberg) [inline monitor.go:113]: Let’s rename this to MonitorEvent to make the name more descriptive.

File: monitor.go [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+	monitorFlagsInitOnce sync.Once
+)
+
+type EventType int
+
+const (
+	EventTypeNewTable   EventType = unix.NFT_MSG_NEWTABLE
+	EventTypeDelTable   EventType = unix.NFT_MSG_DELTABLE
+	EventTypeNewChain   EventType = unix.NFT_MSG_NEWCHAIN
+	EventTypeDELChain   EventType = unix.NFT_MSG_DELCHAIN
```
Comment (stapelberg) [inline monitor.go:101]: This one should be EventTypeDelChain (not uppercase DEL)

**Review by stapelberg — COMMENTED** [review]
Looks good! When you resolve the remaining two comments, I think this is ready to be merged :)

**Review by stapelberg — APPROVED** [review]

**Pull request merged**